torchtune/utils/_activations_offloading/offload_ctx_mgr.py

```diff
@@ -0,0 +1,278 @@
+                            _, ev = self.fwd_stash[k]
+                            self.s0.wait_event(ev)
+                            del self.fwd_stash[k]
+                        else:
+                            break
+
+                    # Sync in, offload, and add an event to sync back later
+                    self.s1.wait_stream(self.s0)
+
+                stream = self.s1 if use_streams else self.s0
+                with torch.cuda.stream(stream):
+                    cpu_tensor = torch.empty_like(
+                        activation,
+                        pin_memory=self.use_pin_memory,
+                        device=torch.device("cpu"),
+                    )
+
+                    cpu_tensor.copy_(activation, non_blocking=True)
+                    self.tracker[tensor_id] = (
+                        cpu_tensor,
+                        True,
+                    )  # True = (in future) modified
+
+                if use_streams:
+                    event = self.s1.record_event()
+
+                    # Stash to keep activation alive til s1 is done
+                    self.fwd_stash[tensor_id] = (activation, event)
+            else:
+                self.tracker[tensor_id] = (
+                    activation,
+                    False,
+                )  # False = not modified, tensor is as is
+
+            return tensor_id
+
+        def unpack_tensor(unpack_tensor_id: int) -> torch.Tensor:
+            # backward pass - we are called with the tensor_id, which
+            # we will use to retrieve the saved/offloaded tensor
+            if self.is_first_backward_call:
+                if use_streams:
+
+                    def wait_and_del_remaining_references() -> None:
+                        for id in [k for k in self.bwd_tensor_stash.keys()]:
+                            event = self.bwd_ev_stash[id]
+                            self.s1.wait_event(event)
+                            del self.bwd_tensor_stash[id]
+
+                    # Register a callback to the end of autograd to clean everything up
+                    torch.autograd.variable.Variable._execution_engine.queue_callback(
+                        wait_and_del_remaining_references
+                    )
+
+                if self.is_first_forward_pass:
+                    self.is_first_forward_pass = False
+                    if self.use_pin_memory:
+                        verify_sufficient_virtual_memory()
+
+                self.is_first_backward_call = False
+                self.is_first_forward_call = True
+
+            assert (
+                unpack_tensor_id in self.tracker
+            ), f"untracked tensor with id {unpack_tensor_id}"
+
+            maybe_gpu_tensor, modified = self.tracker[unpack_tensor_id]
+            if modified:
+                if not use_streams:
+                    gpu_tensor = maybe_gpu_tensor.to(device="cuda", non_blocking=True)
+                    maybe_gpu_tensor = gpu_tensor
+                else:
```

Comment:
More of an organizational thing, but I wonder if it's clearer to just have separate `unpack_tensor_{streams, no_streams}` methods since the logic for the two cases is mostly disjoint anyways.

Reply:
Sure, I can move it out.